cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst

```diff
@@ -0,0 +1,188 @@
+configuration happens by calling the ``configure()`` method of each entity, which creates
+a specification for that entity. These specifications are bundled together to create the
+application.
+
+This results in application, program, and sub-program specifications. In a specification,
+you can set the name, description, resources, programs and sub-programs used, plugins used
+and required. (An example of the latter is a requirement for a JDBC driver.)
+
+Any member variables created in the ``configure()`` methods are available only at
+deployment. No dataset operations are possible, as there is no access to datasets or
+transactions.
+
+**When an application is run,** each program of an application (and any sub-programs) is
+executed by calling first ``initilaize()`` and (eventually) ``destroy()`` for each program
+and sub-program. In these methods, dataset operations and transactions are allowed.
+
+The following of the program lifecycle is a combination of CDAP's conventions and the implmentation
+of specific interfaces. They can be summarised as:
+
+- At deployment time:
+
+  - ``configure()`` By convention, all CDAP applications and programs have this method.
+    It produces an immutable program specification.
+    
+- At runtime:
+
+  - ``initialize()`` A requirement of the ``ProgramLifecycle`` interface.
+  - ``destroy()`` A requirement of the ``ProgramLifecycle`` interface.
+
+The ``initialize()`` method is called once, at the start of the program. The ``destroy()``
+method is called once, at the end of program before it is shutdown. If there is any
+cleanup required, it can be specified in this method.
+
+Flows and services do not have an ``initialize()`` because they have a sub-program (flowlets
+for a flow, service handlers for a service) which has an ``initialize()`` method instead.
+
+Note that the instance of the object called at deployment **is not the same** instance of
+the object called at runtime. Because the result of the deployment stage is an immutable
+program specification, any local member variables set during deployment will not be
+available during runtime. This behavior can cause unexpected null-pointer exceptions. The
+solution is instead to set these as properties in a configuration object, which is
+available at runtime, as in the examples on :ref:`initializing instance fields
+<best-practices-initializing>`.
+
+Transactions
+============
+The relationship between transactions and lifecycle depends on the method involved:
+
+- ``configure()`` No transactions
+
+- ``initialize()`` Inside a transaction
+
+- ``destroy()`` Inside a transaction
+
+The exception to this are :ref:`Workers <workers-datasets>`, which are run inside their own transaction.
+See :ref:`workers and datasets <workers-datasets>` for details.
+
+Details on transactions in these methods is covered in the section on :ref:`using the
+transaction system in programs <transaction-system-using-in-programs>`.
+
+Program Types
```

Comment:
Because you have not mentioned any of the abstract program classes yet, they seem to come out of nowhere in the table that follows. I would add a small subsection on abstract programs: 

For convenience, most program types have a corresponding abstract program class. It is recommended to always extend abstract class instead of implementing the program interface. The abstract classes mainly provide:
- proxy methods for the program configurer's methods. 
- an initialize method that stores the program context in a class member and makes it available via getContext()
- a destroy method that does nothing.